### DISCUSSION TITLE
Arbitrary Inputs for DemoModel() Class

### DISCUSSION THREAD

**Discussion by solidkon** [mupif/mupif#56] [Q&A]
Hi,

I'm trying to define the inputs for the DemoModel() Class provided at : https://gist.github.com/stanislavsulc/5944f1f10d50f37ece05690b8a4d9da7

Practically, the grain components (different type of molecules inside a grain) should be defined by the platform user. This implies that in previous steps of the workflow, the HeavyStruct objects storing the information about the atoms/molecules/grains have already been created. A general Molecular Dynamics simulation should be able to run, independently of how many different user-selected molecules will be considered. Currently I don't know how to implement this in the DemoModel() without initializing the Class via a modified MD (MetaData) dictionary. Am I proceeding the right way or there is a better way of doing this?

**Comment by eudoxos** [comment]
We discussed briefly with Stanislav and need some help understanding what the issue is. We are both monkeys in molecular dynamics and understand only what an average 10-years old child would. Perhaps some schematics of what the model inputs/outputs would be helpful, not sure.

Do you need an arbitrary number of inputs (which we already discussed) of the HeavyStruct type? Do you need to specify that the input is only a HeavyStruct with only certain type of data?

**Reply by solidkon** [reply]
Dear Eudoxos. I think that a discussion between our teams will most certainly clarify everything about this model API. We already have some ideas on how to handle this but we would most certainly need your expertise in handling the model properly. Here what is arbitrary is the number of hdf5 containers created for each type of let's say molecule species being part of a molecular assembly. The easiest way of putting this into a data structure would be to have as an input for the model a list of grains:

`grains = ['g1.h5', 'g2.h5', 'g3.h5', ...]`

Thanks to your implementation of the containers we came up with a special type of a container tailored to our needs, which we call (for the moment) grainStar or ReferenceGrain. This container has as parts 3 MoleculeStates corresponding to 3 possible chargedStates ('neutral', 'asnion', 'cation') for a particular molecule species. This ReferenceGrain contains the  values for each molecule in the molecular assembly (all these properties that are repeated inside the assembly). So actually for this model API we need to somehow handle an arbitrary number of heavystructs. Does this make things a bit clearer?

However, the number of different species inside a molecular assembly is never too large. Specifically for the Musicode user cases a super complex system would be to have a ternary system with dopants inside, thus having a realistic molecular assembly of 5 different molecule species. So the number of possible heavystructs is 6 (5 ReferenceGrains and 1 describing the assembly)

**Reply by solidkon** [reply]
Can you please guide me on how to define:

1. a list of grain hdf5 containers
2. a list of Pyro5 files

as Inputs for the DemoModel() class?

**Reply by eudoxos** [reply]
It is always a `HeavyStruct` (which owns the backing HDF5 container) which will be an input to the model. Let me check with Stanislav how to approach this. There are issues with data validation and the web interface if there are arbitrary lists. How about having 6 parameters to the model (6 HeavyStructs) which would optionally be empty, would that serve as a good solution until we figure out how to handle lists, if it is doable at all?

**Reply by solidkon** [reply]
yes eudoxos this would be a good temporary solution. I understand how the inputs  work in MetaData of the __init__(). If only you could provide me with the MD with 6 optionally empty hdf5 containers and accordingly 6 Pyro5 files, I can proceed with the rest of the preliminary Model implementation

**Reply by stanislavsulc** [reply]
The "Inputs" item in metadata for 6 grains and 6 input files is here:
```
"Inputs": [
    {"Type": "mupif.HeavyStruct", "Type_ID": "mupif.DataID.ID_GrainState", "Name": "Grain",
    "Description": "Input grain", "Required": False, "Set_at": "timestep", "Units": "",
    "Obj_ID": ["1", "2", "3", "4", "5", "6"]},
    {"Type": "mupif.PyroFile", "Type_ID": "mupif.DataID.ID_InputFile", "Name": "Input_file",
    "Description": "Input file", "Required": False, "Set_at": "initialization", "Units": "",
    "Obj_ID": ["1", "2", "3", "4", "5", "6"]}
]
```
Note that the "Obj_ID" can be a single string or a list of strings, which is used here. You can analogically add a single input item.
These items have the "Required" property set to False, which enables them to be unset. The model implementation itself must check if the input is provided or not.
